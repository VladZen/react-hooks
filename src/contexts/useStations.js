import { useState, useCallback } from 'react'
import constate from 'constate'

import Api from '@/api/Stations'

const useStations = () => {
  const [stationsAreLoading, setLoading] = useState(false)
  const [stations, setStations] = useState([])

  const fetchStations = useCallback(async () => {
    setLoading(true)
    try {
      const { data: stations } = await Api.getList()
      setStations(stations)
    } catch (e) {
      console.log(e)
    } finally {
      setLoading(false)
    }
  }, [])

  const getStation = async (id) => {
    if (!stations.length) await fetchStations()
    return stations.find(({ station_ID }) => station_ID === id)
  }

  return {
    stations,
    getStation,
    fetchStations,
    stationsAreLoading
  }
}

// 2️⃣ Wrap your hook with the constate factory
const [StationsProvider, useStationsContext] = constate(useStations)

export { StationsProvider, useStationsContext }
