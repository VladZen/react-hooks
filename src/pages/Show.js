import { useMemo } from 'react'

import { Redirect, useParams } from 'react-router-dom'

import { useStationsContext } from '@/contexts/useStations'

import Header from '@/components/Header'
import StationView from '@/components/StationView'
import Back from '@/components/BackButton'
import Loading from '@/components/Loading'

const Show = () => {
  const {
    isLoading,
    stations
  } = useStationsContext()

  const {
    stationID
  } = useParams()

  const station = useMemo(
    () => stations.find(({ station_ID }) => station_ID === parseInt(stationID)),
    [stations, stationID]
  )

  return (
    <div className='layout'>
      <Loading trigger={isLoading}>
        {
          !station
            ? <Redirect to='/' />
            : <>
              <Header heading={ station.name }>
                <Back to='/' />
              </Header>

              <main className='layout__main'>
                <StationView station={ station } />
              </main>
            </>
        }
      </Loading>
    </div>
  )
}

export default Show
