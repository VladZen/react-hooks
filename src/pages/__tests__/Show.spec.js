import { render } from '@testing-library/react'
import { useParams, MemoryRouter } from 'react-router-dom'
import { useStationsContext } from '@/contexts/useStations'
import Show from '@/pages/Show'

jest.mock('@/contexts/useStations', () => ({
  ...jest.requireActual('@/contexts/useStations'), // use actual for all non-hook parts
  useStationsContext: jest.fn()
}))

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useParams: jest.fn()
}))

const stations = [
  {
    'station_ID': 103,
    'custom_evse_id': null,
    'location_ID': 103,
    'seller_ID': 27,
    'name': 'Endesa 3',
    'connected': 1,
    'position': '60.868623,26.702901',
    'available': 0,
    'lastconnect': '2019-03-26T06:46:32.000Z',
    'roaming_identifier_cpo': 'FI*001'
  },
  {
    'station_ID': 104,
    'custom_evse_id': null,
    'location_ID': 104,
    'seller_ID': 27,
    'name': 'Endesa 4',
    'connected': 1,
    'position': '60.868623,26.702901',
    'available': 1,
    'lastconnect': '2018-03-15T11:25:47.000Z',
    'roaming_identifier_cpo': 'FI*001'
  }
]

describe('Show page ', () => {
  describe('rendering', () => {
    it('loading', () => {
      useStationsContext.mockReturnValueOnce({
        isLoading: true,
        stations: []
      })

      useParams.mockReturnValueOnce({
        stationID: stations[0].station_ID
      })

      const { container } = render(
        <Show />
      )

      expect(container).toMatchSnapshot()
    })

    it('loaded', () => {
      useStationsContext.mockReturnValueOnce({
        isLoading: false,
        stations
      })

      useParams.mockReturnValueOnce({
        stationID: stations[0].station_ID
      })

      const { container } = render(
        <MemoryRouter>
          <Show />
        </MemoryRouter>
      )

      expect(container).toMatchSnapshot()
    })

    it('shows 404 if id is not found', () => {
      useStationsContext.mockReturnValueOnce({
        isLoading: false,
        stations
      })

      useParams.mockReturnValueOnce({
        stationID: 5000
      })

      const { container } = render(
        <MemoryRouter>
          <Show />
        </MemoryRouter>
      )

      expect(container).toMatchSnapshot()
    })
  })
})

