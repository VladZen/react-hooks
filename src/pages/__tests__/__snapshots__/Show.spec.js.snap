// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show page  rendering loaded 1`] = `
<div>
  <div
    class="layout"
  >
    <header
      class="layout__header"
    >
      <div
        class="layout__header__buttons"
      >
        <a
          class="back-button"
          href="/"
          title="Back to the list"
        >
          <svg
            class="back-button__icon"
          >
            <use
              href="sprite.svg#arrow"
            />
          </svg>
        </a>
      </div>
      <h1
        class="text-ellipsis"
        title="Endesa 3"
      >
        Endesa 3
      </h1>
    </header>
    <main
      class="layout__main"
    >
      <dl
        class="station-attributes"
      >
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Station ID
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            103
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Custom Evse Id
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            Not detected
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Location ID
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            103
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Seller ID
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            27
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Name
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            Endesa 3
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Connected
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            <span
              class="text-positive"
            >
              Yes
            </span>
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Position
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            <a
              class="link-decorated"
              href="http://www.google.com/maps/place/60.868623,26.702901"
              target="blank"
              title="Open in new tab"
            >
              Show on the map
            </a>
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Available
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            <span
              class="text-negative"
            >
              No
            </span>
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Lastconnect
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            3/26/2019, 9:46:32 AM
          </dd>
        </div>
        <div
          class="station-attributes__item"
        >
          <dt
            class="station-attributes__item__label"
          >
            Roaming Identifier Cpo
          </dt>
          <dd
            class="station-attributes__item__value"
          >
            FI*001
          </dd>
        </div>
      </dl>
    </main>
  </div>
</div>
`;

exports[`Show page  rendering loading 1`] = `
<div>
  <div
    class="layout"
  >
    <div
      class="spinner"
    >
      Loading...
    </div>
  </div>
</div>
`;

exports[`Show page  rendering shows 404 if id is not found 1`] = `
<div>
  <div
    class="layout"
  >
    <div
      class="layout"
    >
      <header
        class="layout__header"
      >
        <div
          class="layout__header__buttons"
        >
          <a
            class="back-button"
            href="/"
            title="Back to the list"
          >
            <svg
              class="back-button__icon"
            >
              <use
                href="sprite.svg#arrow"
              />
            </svg>
          </a>
        </div>
        <h1
          class="text-ellipsis"
          title="Page Not Found"
        >
          Page Not Found
        </h1>
      </header>
    </div>
  </div>
</div>
`;
